RUFAS/biophysical/animal/pen.py

```diff
@@ -894,106 +896,42 @@ def formulate_optimized_ration(  # noqa: C901
-            ration_optimizer.handle_failed_constraints(
-                num_attempts=num_attempts,
-                solution=solution,
-                ration_config=ration_config,
-                animal_combination=self.animal_combination,
-                pen_id=self.id,
-                pen_available_feeds=pen_available_feeds,
-                average_nutrient_requirements=self.average_nutrition_requirements,
-                sim_day=9999,
-                info_map=info_map,
-            )
-            # TODO get sim day from RuFaS time, fixed in PR #2381
-        if self.animal_combination == AnimalCombination.LAC_COW:
-            while not solution.success:
-                if self.average_milk_production < AnimalModuleConstants.MINIMUM_AVG_PEN_MILK:
-                    self.om.add_error(
-                        "Milk production too low",
-                        (
-                            f"Check failed_constraint_summary_for_pen_{self.id} to see what caused formulation to fail."
-                            f" Possible solution is to provide additional feed ingredients to "
-                            f"{self.animal_combination.name}."
-                        ),
-                        info_map,
-                    )
-                    raise ValueError(
-                        f"Check failed_constraint_summary_for_pen_{self.id} to see what caused"
-                        f"formulation to fail. Possible solution is to provide additional feed"
-                        f"ingredients to {self.animal_combination.name}."
-                    )
-                could_reduce = self.reduce_milk_production()
-                if not could_reduce:
-                    self.om.add_error(
-                        "Milk production reduced below reduction maximum.",
-                        (
-                            f"Check failed_constraint_summary_for_pen_{self.id} to see what caused formulation to fail."
-                            f" Possible solution is to provide additional feed ingredients to "
-                            f"{self.animal_combination.name}."
-                            f"Also consider increasing the milk reduction maxmimum in input JSON."
-                        ),
-                        info_map,
-                    )
-                    raise ValueError
-                self.set_animal_nutritional_requirements(temperature=temperature, available_feeds=pen_available_feeds)
+        previous_ration = getattr(self, "ration", None)
+        num_attempts = 0
 
-                solution, ration_config = ration_optimizer.attempt_optimization(
-                    pen_average_body_weight=self.average_body_weight,
-                    requirements=self.average_nutrition_requirements,
-                    pen_available_feeds=pen_available_feeds,
+        while True:
+            num_attempts += 1
+            solution, ration_config = self._attempt_formulation(pen_available_feeds, temperature, previous_ration)
```

Comment:
@JoeWaddell or @matthew7838 Can one of you explain a little how this section handles `None` being returned from `self._attempt_formulation()` for the `solution` part of the tuple? And maybe a little about what set of circumstances would lead to the solution being `None` in the first place? Just want to be sure that's all handled and accounted for here.